src/ecs/resources/input.rs

```diff
@@ -433,82 +390,118 @@ impl InputHandler {
 
     /// Assign an axis to an ID value
     ///
     /// This will insert a new axis if no entry for this id exists.
     /// If one does exist this will replace the axis at that id and return it.
-    pub fn insert_axis(&mut self, id: i32, axis: Axis) -> Option<Axis> {
+    pub fn insert_axis(&mut self, id: String, axis: Axis) -> Option<Axis> {
         self.axes.insert(id, axis)
     }
 
     /// Removes an axis, this will return the removed axis if successful.
-    pub fn remove_axis(&mut self, id: i32) -> Option<Axis> {
-        self.axes.remove(&id)
+    pub fn remove_axis<T: AsRef<str>>(&mut self, id: T) -> Option<Axis> {
+        self.axes.remove(id.as_ref())
     }
 
     /// Returns a reference to an axis.
-    pub fn get_axis(&mut self, id: i32) -> Option<&Axis> {
-        self.axes.get(&id)
+    pub fn axis<T: AsRef<str>>(&mut self, id: T) -> Option<&Axis> {
+        self.axes.get(id.as_ref())
     }
 
-    /// Assign an action to an ID value
-    ///
-    /// This will insert a new action if no entry for this id exists.
-    /// If one does exist this will replace the action at that id and return it.
-    pub fn insert_action(&mut self, id: i32, binding: Button) -> Option<Button> {
-        self.actions.insert(id, binding)
+    /// Get's a list of all axes
+    pub fn axes(&self) -> Vec<String> {
+        self.axes.keys().map(|k| k.clone()).collect::<Vec<String>>()
     }
 
-    /// Removes an action, this will return the removed action if successful.
-    pub fn remove_action(&mut self, id: i32) -> Option<Button> {
-        self.actions.remove(&id)
+    /// Add a button to an action.
+    ///
+    /// This will insert a new binding between this action and the button.
+    pub fn insert_action_binding<T: AsRef<str>>(&mut self, id: T, binding: Button) {
+        let mut make_new = false;
+        match self.actions.get_mut(id.as_ref()) {
+            Some(action_bindings) => {
+                if action_bindings.iter().all(|&b| b != binding) {
+                    action_bindings.push(binding);
+                }
+            }
+            None => {
+                make_new = true;
+            }
+        }
+        if make_new {
+            let mut bindings = SmallVec::new();
+            bindings.push(binding);
+            self.actions.insert(id.as_ref().to_string(), bindings);
+        }
     }
 
-    /// Returns a reference to an action's button.
-    pub fn get_action(&mut self, id: i32) -> Option<&Button> {
-        self.actions.get(&id)
+    /// Removes an action binding that was assigned previously.
+    pub fn remove_action_binding<T: AsRef<str>>(&mut self, id: T, binding: Button) {
+        let mut kill_it = false;
+        if let Some(action_bindings) = self.actions.get_mut(id.as_ref()) {
+            let index = action_bindings.iter().position(|&b| b == binding);
+            if let Some(index) = index {
+                action_bindings.swap_remove(index);
+            }
+            if action_bindings.len() == 0 {
```

Comment:
Let's use late binding here: Change the line at the beginning of the method to `let kill_it;` and change this line to `kill_it = action_bindings.is_empty()`.